Change interaction modal in web UI

Gargron opened the pull request:
This is a rework of the remote interaction modal. Instead of suggesting people to copy-paste the permalink into their favourite Mastodon app or interface, we let them enter their home server's domain (the input also accepts full username/domain handles and URLs), we save this in local storage for future visits, and redirect them to their home server's view of said profile or post.

https://github.com/mastodon/mastodon/assets/184731/430cccfa-5adf-4865-bdc2-86ed09390b21

___

Fixes MAS-115

Gargron committed:
Change interaction modal in web UI

trwnh reviewed (changes requested):
can this be standardized instead of hardcoded to whatever mastodon does? why don't we use webfinger?

### proposal 1: use the existing `proxyUrl` as defined by activitypub

**new link relation**: `https://www.w3.org/ns/activitystreams#proxyUrl`

as defined in activitypub https://www.w3.org/TR/activitypub/#proxyUrl

> Endpoint URI so this actor's clients may access remote ActivityStreams objects which require authentication to access. To use this endpoint, the client posts an `x-www-form-urlencoded` `id` parameter with the value being the `id` of the requested ActivityStreams object.

```json
{
  "subject": "...",
  "aliases": [...],
  "links": [
    {
      "rel": "https://www.w3.org/ns/activitystreams#proxyUrl",
      "href": "https://mastodon.example/authorize_interaction"
    }
  ]
}
```

### proposal 2: extend this behavior to allow HTTP GET

the above definition uses HTTP POST and a form data parameter. it should also be possible to call this endpoint by getting it with an `id` query parameter. a public representation MAY be returned if the request is unauthenticated. if the `Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"` header is provided, the ActivityStreams document is provided.

```http
GET /authorize_interaction?id=... HTTP/1.1
```

Gargron commented:
Hi @trwnh this is a first pass to get it working. My thinking is ideally we'd use a `/.well-known/` URL since there's already prior art with `/.well-known/change-password`. I don't necessarily want to bind this UX to webfinger; I'm not sure if in practice username@domain is easier or harder to remember than just domain but it feels to me the less the user has to type the better (plus you don't have to reveal who exactly you are).

trwnh commented:
that's fair. i think in some cases, you might be able to get away with just the domain, which might possibly have its own JRD representation in webfinger when querying just for the domain. or you could have a new .well-known endpoint like you said. but there's some conceptual precedent to having this be a form of `proxyUrl` as defined by activitypub. webfinger + proxyUrl feels like it's not straying too far from standards, imo

joenepraat commented:
Looking good, but I have a remark about 'Take me home'. I don't think it's clear enough that button will do the requested action. I advise to use direct terminology, like 'Follow' and 'Reply'. Or if there is a redirect to the users server; 'Follow and go home' and 'Reply and go home'.

j-f1 commented:
Maybe something like “Follow on my server” or “Follow from my server?”

Gargron commented:
It can't execute the actual action from this screen. That's why I chose "Take me home".

ThisIsMissEm commented on app/controllers/api/v1/peers/search_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  private
+
+  def set_domains
+    return if !Chewy.enabled? || params[:q].blank?
```
Could we fallback to just a fuzzy match in PostgreSQL? Perhaps using the "matches_domain" scope?

joenepraat commented:
> It can't execute the actual action from this screen. That's why I chose "Take me home".

Aha clear. I thought it was something FediAct does, but it redirects to your home server where  the user can take the action. Thanks for the explanation.

ThisIsMissEm commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
+    if (segments.length === 2) {
+      return segments[1];
+    } else {
+      return '';
```
This'll cause problems if/when domain-only accounts are a thing; perhaps it'd be better to perform an account search here?

ClearlyClaire replied:
There's another more immediate issue: this doesn't support situations where `LOCAL_DOMAIN` ≠ `WEB_DOMAIN`. Indeed, if `@foo@bar.name` is served by `https://social.bar.name/users/foo`, this whole flow redirects to `https://social.bar.name/authorize_interaction?uri=…`, which requires an extra redirect, while 3.x's interaction modal handled that case just fine.

ThisIsMissEm commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
+    }
+
+    if (localStorage) {
+      localStorage.setItem(PERSISTENCE_KEY, domain);
```
Would it be worth making this an array value so I can have multiple domains remembered? E.g., I have accounts on Mastodon, Pixelfed, and Calckey, I'd wamt to pick between each here

ThisIsMissEm commented on config/routes/api.rb:
```diff
@@ -121,6 +121,10 @@
     end
 
+    namespace :peers do
+      get :search, to: 'search#index'
```
Shouldn't search be a POST request?

Gargron replied:
I checked and other search APIs we have are GET

ThisIsMissEm reviewed (commented):
Looks good to me, just a few comments. 

Also, it might be worth us adding a proper autocomplete component to the stack as they can be rather difficult to get all the interactions correct (e.g., something like downshift)

ThisIsMissEm commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
 
   state = {
-    copied: false,
+    value: localStorage ? (localStorage.getItem(PERSISTENCE_KEY) || '') : '',
```
What happens if the stored value is corrupted? Should we validate it after loading from localstorage?

ClearlyClaire commented on app/javascript/mastodon/features/compose/components/search.jsx:
```diff
@@ -135,10 +135,6 @@ class Search extends React.PureComponent {
     this.setState({ expanded: false, selectedOption: -1 });
   };
 
-  findTarget = () => {
-    return this.searchForm;
-  };
```
Unrelated but good.

ClearlyClaire commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
+      localStorage.setItem(PERSISTENCE_KEY, domain);
+    }
+
+    const redirectUrl = new URL(`https://${domain}/authorize_interaction?uri=${encodeURIComponent(resourceUrl)}`);
```
This will only work with Mastodon, not Pleroma, Akkoma, Misskey, Calckey, or PixelFed, while some of these do support the old flow.

People using any of these services will have either a 404, or in some case, end up on the profile of an account called `authorize_interaction` if it exists.

Gargron replied:
If I use a new .well-known URL now, it won't work for most people on release.

ClearlyClaire replied:
That's true… that and unambiguous support for configurations where `LOCAL_DOMAIN` ≠ `WEB_DOMAIN` are the reasons why I think we should rather ask for the handle rather than the server, this would allow us to use the already well-supported flow.

ClearlyClaire replied:
It worked. And protocol-wise, that's not “just to get the second part”. Multiple AP software can theoretically run on the same domain. And more long-term thinking, if we ever move to DIDs or even just to have multi-tenancy and e.g. mail providers allowing you to point to your Mastodon instance to have the same fediverse handle as your mail address, we can't rely solely on the domain.

Gargron replied:
I think it's much easier to type a domain with autocomplete suggestions than to type your entire username (my instinct is that the autocomplete will play a major role in how user-friendly this feature will be perceived), I also don't think it's necessarily good to require people to give away their identity for this. Most apps require only the domain for the login. I would love to deploy this and observe how people get along with it in practice.

trwnh replied:
could we do this via webfinger against the bare domain name as resolved via the webfinger host? i know conceptually it is only 100% unambiguous if given either an acct: or an actor id, because then you could define something like what i described above with proxyUrl. but if the important bit is specifically "put in a domain name only" then you could still get there by having the domain resolve via either webfinger or activitypub. (claire raises a good point about the differing domains and also different softwares on the same domain, which i'm sure cause problems in other parts of the code as well, but one thing at a time)

ClearlyClaire replied:
> I think it's much easier to type a domain with autocomplete suggestions than to type your entire username (my instinct is that the autocomplete will play a major role in how user-friendly this feature will be perceived)

I'm not completely sure. And I think autocomplete on handle could even be more useful, but it's much less likely to already know about your account, so it's definitely not a clear win either.

That being said, we can support both: if it looks like a handle, autocomplete on account search, otherwise, autocomplete on domain. You could even imagine autocompleting the domain part if no user matches.

> I also don't think it's necessarily good to require people to give away their identity for this

That is a very good point, although the modal's point is for you to interact with the resource from your remote account, so in most cases they'd give away their identity on the next step.

> Most apps require only the domain for the login.

True, and it can already cause quite some confusion when `WEB_DOMAIN` ≠ `LOCAL_DOMAIN`.

> I would love to deploy this and observe how people get along with it in practice.

I agree, but I would also like not committing to something that is neither future-proof nor compatible with the non-Mastodon fediverse.

> could we do this via webfinger against the bare domain name as resolved via the webfinger host?

I don't think that's currently supported, so that would present most of the same drawbacks as the current PR while also breaking compatibility with current Mastodon :/

Gargron commented:
I've changed it to use `/.well-known/proxy` and added a redirect. This won't work for any existing Mastodon prior to this patch or other software but at least it doesn't hardcode a Mastodon path.

trwnh commented:
why `/.well-known/proxy` specifically? that's not an [existing well-known uri](https://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml), and if we're going to use a new uri entirely, then it would make sense to actually define it as a standard. `proxy` sounds too generic on its own. i really think this makes sense in webfinger, or at least something allowing for link relations (such that a rel-value of `https://www.w3.org/ns/activitystreams#proxyUrl` can unambiguously point toward an endpoint providing the expected semantics as described above).

trwnh commented:
> What about storing the url in the metadata from `.well-known/nodeinfo` ? Sure, it'd require 3 requests to resolve but might at least then but standardisable?

webfinger would be 2 requests: one to get the JRD, one to follow the link. the important thing imo is what to do after following the link. and you can avoid asking for the full user@domain if you instead ask for the domain and assume https. just make webfinger handle requests for https uris for the LOCAL_DOMAIN, WEB_DOMAIN and any ALTERNATE_DOMAINS maybe? and on the endpoint itself just make the request against the `href` using the standardized `id` parameter. i'm open to better suggestions but i think this covers everything you'd want.

Gargron commented:
It feels like none of the good approaches are backwards-compatible. Webfinger won't respond to empty `resource` requests, and we don't have a handler for "just domain" requests. We could try to guess a user, or use some known user, but it won't work for every domain. Worth noting that `http://ostatus.org/schema/1.0/subscribe` in the webfinger is not part of any standard and isn't even entirely accurate naming-wise anymore ("subscribe"?), not to mention the expired ostatus.org domain which I've received a few complaints about. In my view the most practical approach *would've* been to hardcode-redirect to `/authorize_interaction` to get this UI out there and functional immediately, then implement whatever API is necessary on the other end, and roll out a new version using it afterwards 🤷‍♂️

trwnh commented:
> Webfinger won't respond to empty resource requests

how about `https://mastodon.social/.well-known/webfinger?resource=https://mastodon.social` is what i meant. as a fallback, you can try the old hardcoded path, but i think some effort should be made to try defining and looking up a standardized endpoint which is mostly what i'm trying to describe.

basically i'm proposing logic like so:

- take the user's input
  - if it's a user@domain:
    - try webfinger for that `acct:user@domain`
    - get the rel=as:proxyUrl link.
      - if not found, continue to next step
  - if it's a bare domain:
    - try webfinger for that `https://domain`
    - get the rel=as:proxyUrl link.
      - if not found:
        - try `/authorize_interaction`
          - if not found, abort

and on the `/authorize_interaction` side:

- change the handler to accept `id` and not just `uri`
- change the handler to accept POST in addition to GET

and on the webfinger side:

- respond to requests for `https://WEB_DOMAIN`, `https://LOCAL_DOMAIN`, and any `https://ALTERNATE_DOMAINS`
- include a link with rel=as:proxyUrl on both users and on this new domain JRD

and on the activitypub side:

- expose `/authorize_interaction` via `endpoints.proxyUrl`

this should be backwards compatible with at least mastodon due to the fallback logic, and it should also prevent outcomes where other software is forced to hardcode mastodon-specific logic (unless they also want to make use of the fallback)

ClearlyClaire commented:
> > Webfinger won't respond to empty resource requests
> 
> how about `https://mastodon.social/.well-known/webfinger?resource=https://mastodon.social` is what i meant. as a fallback, you can try the old hardcoded path, but i think some effort should be made to try defining and looking up a standardized endpoint which is mostly what i'm trying to describe.
> 
> basically i'm proposing logic like so:
> 
>     * take the user's input
>       
>       * if it's a `user@domain`:
>         
>         * try webfinger for that `acct:user@domain`
>         * get the `rel: https://www.w3.org/ns/activitystreams#proxyUrl` link.
>           
>           * if not found, continue to next step
>       * if it's a bare `domain`:
>         
>         * try webfinger for that `https://domain`
>         * get the `rel: https://www.w3.org/ns/activitystreams#proxyUrl` link.
>           
>           * if not found:
>             
>             * try `/authorize_interaction`
>               
>               * if not found, abort
> 
> 
> and on the `/authorize_interaction` side:
> 
>     * change the handler to accept `id` and not just `uri`
> 
>     * change the handler to accept POST in addition to GET
> 
> 
> and on the webfinger side:
> 
>     * respond to requests for `https://WEB_DOMAIN`, `https://LOCAL_DOMAIN`, and any `https://ALTERNATE_DOMAINS`
> 
>     * include a link with `rel: https://www.w3.org/ns/activitystreams#proxyUrl` on both users and on this new domain JRD
> 
> 
> and on the activitypub side:
> 
>     * expose `/authorize_interaction` via `endpoints.proxyUrl`
> 
> 
> this should be backwards compatible with at least mastodon due to the fallback logic, and it should also prevent outcomes where other software is forced to hardcode mastodon-specific logic (unless they also want to make use of the fallback). you could even add a third case where if a user inputs an https profile url.

That sounds very sensible, although I would add a check for `http://ostatus.org/schema/1.0/subscribe` along `https://www.w3.org/ns/activitystreams#proxyUrl` for backward compatibility when querying `user@account`.

trwnh commented:
one other thing i forgot to describe: instead of aborting completely you could have the current "copy paste the link into your resolver / search box" dialog be shown as the failure state. something like "sorry, we couldn't find a way to automatically redirect you" of sorts

SoniEx2 commented:
why not `/.well-known/protocol-handler?target=web%2Bap://...`?

SoniEx2 commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -1,103 +1,251 @@
+      localStorage.setItem(PERSISTENCE_KEY, domain);
+    }
+
+    const redirectUrl = new URL(`https://${domain}/.well-known/proxy?uri=${encodeURIComponent(resourceUrl)}`);
```
would you mind using `protocol-handler` instead of `proxy` here?

trwnh commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -1,103 +1,261 @@
+    this.setState({ isSubmitting: true });
+
+    api().get(`https://${domain}/.well-known/webfinger`, { params: { resource: domain } }).then(({ data }) => {
+      const template = data.links.find(link => link.rel === 'http://ostatus.org/schema/1.0/subscribe')?.template;
```
does this bit only pick up on the ostatus rel and the `template` value? i can't find a definition for `template`, and it's not in RFC 7033, so i'd like to know what you think about the proposal for extending as:proxyUrl

Gargron replied:
It feels weird to use the path to a JSON-LD property as a `rel` value. Maybe we could use something from [here](https://microformats.org/wiki/existing-rel-values), if we're adding new properties anyway? There seems to even be something called `respond-proxy`, which sounds similar to our use case.

trwnh replied:
the iri for proxyUrl is used to define semantics, and these semantics 90% match what is going on here -- you just need to handle GET in addition to POST (or in mastodon's terms, add support for POST in addition to GET, and the `id` parameter instead of `uri`). it's not weird to use a semantic iri as a rel-value, and arguably if you consider RDF then using it as a property creates a qualified relation anyway, so it's the same thing.

trwnh commented on app/javascript/mastodon/features/interaction_modal/index.jsx:
```diff
@@ -146,22 +304,12 @@ class InteractionModal extends PureComponent {
 
-          <div className='interaction-modal__choices__choice'>
-            <h3><FormattedMessage id='interaction_modal.on_another_server' defaultMessage='On a different server' /></h3>
-            <p><FormattedMessage id='interaction_modal.other_server_instructions' defaultMessage='Copy and paste this URL into the search field of your favourite Mastodon app or the web interface of your Mastodon server.' /></p>
```
can this be kept as a last-resort failure state? in case the webfinger fails and also the fallback url fails, this is useful information to have.

ClearlyClaire commented:
Seems like the PR was not updated with the suggestion to webfinger for `acct:@username@domain` to support `WEB_DOMAIN` ≠ `LOCAL_DOMAIN` (as well as the more generalized multi-tenancy potential of that) and existing fediverse software.

Furthermore, the text “Tip: It's the second half of your username” is misleading when `WEB_DOMAIN` ≠ `LOCAL_DOMAIN`, and suggests that the feature should just work with full usernames instead of asking that work from the user.
